### PR TITLE
hubble-proxy: add initial skeleton

### PR DESCRIPTION
**Pull request by rolinh**
This commit adds code to create a binary, `hubble-proxy`, for the new
hubble proxy. This component is a new component that will be required
for hubble's multi-node feature.

Notably, the introduction of hubble proxy  will:

 - Allow sharing the hubble API externally via load balancer or ingress
   rules.
 - Give flexibility in terms of scheduling and resource allocation.
 - Make it easy to discover from Hubble UI (k8s service) or other
   clients such as the hubble CLI.
 - Make it potentially easier to implement API rate limiting to protect
   the cluster.
 - Decouple security from Cilium (CNPs can be applied).

The hubble proxy will talk directly with the hubble API on the cilium
agent running on the same node. This will allow initial hubble peers
discovery but also to monitor for peers joining/leaving the cluster or
to get information about which peer to connect to to get flows based on
the request (typically for long running requests). The hubble proxy will
connect to hubble peers in order to get relevant flows according to the
flow request to process.

Ref: https://github.com/cilium/hubble/issues/89

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/10545)

### CODE REVIEW

**Comment by Rolinh** [comment]
test-me-please

File: hubble-proxy/Makefile
Comment (gandro) [inline hubble-proxy/Makefile]: I wonder if it would make sense to use Cilium's versioning infrastructure here instead:
https://github.com/cilium/cilium/blob/6d29bba62d7cef9d340848d360da2effc09f2485/Makefile.defs#L34

I think we probably also want to pull in all the other Makefile variables from `../Makefile.defs` so we end up building all binaries with the same `$GOBUILD` etc. See for example the Makefile of the Cilium CLI: https://github.com/cilium/cilium/blob/6d29bba62d7cef9d340848d360da2effc09f2485/cilium/Makefile

Comment (rolinh) [reply hubble-proxy/Makefile]: I thought about this as well but in the end, although `hubble-proxy` lives in `cilium`'s repository (primarily to make integration testing easier), it's entirely independent from `cilium` as it will only interact with the hubble API and will run in an entirely separate container and even pod.
I'm also OK to make use of variables defined in `../Makefile.defs` but I'm not convinced yet that it's the better way to do it.

File: hubble-proxy/cmd/completion/completion.go
Comment (joestringer) [inline hubble-proxy/cmd/completion/completion.go:1]: nit: I think we could just use SPDX here to trim down the license boilerplate for these new files?

File: hubble-proxy/cmd/completion/completion.go
Comment (joestringer) [inline hubble-proxy/cmd/completion/completion.go:43]: I noticed there's some subtle differences between the completion code here and the ones in `cmd/root.go`, is it reasonable to find a way to share this?

File: hubble-proxy/Makefile
Comment (joestringer) [reply hubble-proxy/Makefile]: Seems like there's two questions here:

* Should hubble-proxy use Cilium `Makefile` infrastructure?

My gut feeling here is that creating a new module, developers naturally want to create the simplest possible `Makefile`. Keeps things simple and they just work and it's easier to understand. Over time you need more and more from the make infrastructure that eventually motivates just consuming the same buildchain as the rest of the project. If this module lives in the Cilium project long-term, I would expect this step inevitable; you either reimplement the wheel or reuse it. There's a range of common functionality like avoiding pitfalls around shell invocation by setting the right shell options, to allowing users to configure their container runtime CLI / Go binary to consuming the same `$GOBUILD` flags as the rest of the build when running/testing. Maybe it's not needed today but I expect it's necessary one day.

* What's the intended semantic version handling for `hubble-proxy`?

For this, we can either tie it into Cilium semver or not. Seems like your answer here is expressing desire to allow a separate release / distribution flow for hubble-proxy, so detaching the semver would help with this. Just be aware that if this tool is sitting in the tree primarily for easier integration testing.. the version that will be tested there will be Cilium in the tree + hubble-proxy in the tree; compatibility with other versions, API breakage, etc. will be up to developers to manage and figure out how to continuously integrate.

**Review by joestringer — COMMENTED** [review]
IMO it'd be helpful if the commit and PR message provided a one-liner of the intended scope of this component. Even something simple like "hubble-proxy is a new component to safely expose the Hubble API to remote peers".

Please also extend the `CODEOWNERS` file with a group responsible for owning this module. Feel free to reach out to @aanm or I if you think we need a new team in the cilium organisation to handle reviews and code ownership.

**Comment by joestringer** [comment]
Apparently on the `CODEOWNERS` front, the relevant codeowners were pulled in somehow, even though GitHub also pulled in janitors (which is why I stepped in), and even though I don't see these paths listed either in the existing codeowners file or any changes in this PR to the codeowners file..?

At least right now, GitHub is telling me that I'm a codeowner for all of the files in the review page, which suggests the janitor will be pulled in for any subsequent PRs.

File: hubble-proxy/cmd/completion/completion.go
Comment (rolinh) [reply hubble-proxy/cmd/completion/completion.go:43]: Ah, yes it's because when one runs completion with a wrong argument (eg `hubble-proxy completion foo`), then the copyright header was being written anyway before the error was printed out and it felt weird. I can create a PR to fix this for cilium's cli as well.

**Comment by Rolinh** [comment]
> Apparently on the `CODEOWNERS` front, the relevant codeowners were pulled in somehow, even though GitHub also pulled in janitors (which is why I stepped in), and even though I don't see these paths listed either in the existing codeowners file or any changes in this PR to the codeowners file..?

The relevant codeowners were pulled in because `pkg/hubble` is touched by this PR. But indeed, I should update the `CODEOWNERS` for the new `hubble-proxy` directory.

File: hubble-proxy/Makefile
Comment (rolinh) [reply hubble-proxy/Makefile]: >     * Should hubble-proxy use Cilium `Makefile` infrastructure?

For the time being, I don't think it's necessary. By including `Makefile.defs`, this also defines variables that are non-relevant or even confusing for `hubble-proxy` (eg: `VERSION` as the `hubble-proxy` version may not be tied to the cilium version).

>     * What's the intended semantic version handling for `hubble-proxy`?
> 
> For this, we can either tie it into Cilium semver or not. Seems like your answer here is expressing desire to allow a separate release / distribution flow for hubble-proxy, so detaching the semver would help with this. Just be aware that if this tool is sitting in the tree primarily for easier integration testing.. the version that will be tested there will be Cilium in the tree + hubble-proxy in the tree; compatibility with other versions, API breakage, etc. will be up to developers to manage and figure out how to continuously integrate.

For the time being, this is not defined. Hubble proxy releases could be made independently of cilium as long as hubble's grpc interface remains compatible.

**Comment by Rolinh** [comment]
> Please also extend the `CODEOWNERS`

Done.

File: hubble-proxy/cmd/completion/completion.go
Comment (rolinh) [reply hubble-proxy/cmd/completion/completion.go:43]: See #10558

**Comment by Rolinh** [comment]
test-me-please

File: hubble-proxy/cmd/completion/completion.go
Comment (joestringer) [reply hubble-proxy/cmd/completion/completion.go:43]: Thanks for the fix. Can we refactor / share this or is that painful for some reason (eg the way the viper context is passed around or something?)

**Review by joestringer — APPROVED** [review]
The new description is really helpful, thanks!

I have no general objections from janitor perspective, but i'll defer to @cilium/hubble to review and resolve any outstanding questions.

File: hubble-proxy/Makefile
Comment (glibsm) [reply hubble-proxy/Makefile]: The proxy releases (github tags) can not be made separate from cilium, as the trees are tied together. We can push quay images, but that's about it.

Comment (glibsm) [reply hubble-proxy/Makefile]: Unless there are blockers from us using the existing makefiles, I'd get that straight from the start. Hubble and cilium often live in two completely separate worlds with their own structure due to the team allocation. It makes sense to bridge that a little bit given the current interweave between the two

File: hubble-proxy/cmd/context/context.go
Comment (glibsm) [inline hubble-proxy/cmd/context/context.go]: im' guessing this is probably something ported from cilium, but `context.Context` has a way to pass objects around as well, without the need for a completely separate function signature to pass some additional context structure.

https://golang.org/pkg/context/#WithValue

File: hubble-proxy/cmd/context/context.go
Comment (glibsm) [inline hubble-proxy/cmd/context/context.go]: all this feels like extraneous boilerplate currently not needed or used. hubble proxy is not meant as a binary to be ran on a local laptop, thus probably doesn't need the whole dotfile charade

**Review by michi-covalent — APPROVED** [review]

File: pkg/hubble/proxy/version.go
Comment (aanm) [inline pkg/hubble/proxy/version.go]: If you plan to change this when linking this needs to be a `var`.

File: hubble-proxy/cmd/context/context.go
Comment (rolinh) [reply hubble-proxy/cmd/context/context.go]: So how should configuration be done? By using environment variables only? CLI switches ?

File: hubble-proxy/cmd/context/context.go
Comment (rolinh) [reply hubble-proxy/cmd/context/context.go]: It is not something ported from cilium. The main purpose of this command context is to avoid global vars for the logger and config object to allow for easier testing.

If you insist, I can embed this information into  `context.Context` as you suggest but this will come at the cost of more boilerplate code (to handle `interface{}` type conversion, defining own type for keys as it is a requirement, etc) and decreased visibility (it's not obvious by reading a function signature that the context parameter is expected to carry special values).

File: hubble-proxy/Makefile
Comment (rolinh) [reply hubble-proxy/Makefile]: Update the PR to use cilium's make definitions as we as version.

File: hubble-proxy/cmd/context/context.go
Comment (glibsm) [reply hubble-proxy/cmd/context/context.go]: Context bag is meant for things like rpc/http servers that have to do things over and over. The cmd is bootstrapped once, why do we need context at all?

File: hubble-proxy/cmd/context/context.go
Comment (glibsm) [reply hubble-proxy/cmd/context/context.go]: Presumably this binary is only ever launched from the k8s yaml configuration, so all knobs and switches should be CLI options

**Comment by Rolinh** [comment]
test-me-please

**Comment by Rolinh** [comment]
test-me-please

File: hubble-proxy/Makefile
Comment (errordeveloper) [inline hubble-proxy/Makefile]: This variable unused

**Comment by Rolinh** [comment]
test-me-please

File: hubble-proxy/Makefile
Comment (glibsm) [inline hubble-proxy/Makefile]: proxy probably doesn't rely on all that, right?

File: hubble-proxy/cmd/root.go
Comment (glibsm) [inline hubble-proxy/cmd/root.go]: this feels unnecessary, but more to come in future PRs i'm sure

File: hubble-proxy/cmd/root.go
Comment (glibsm) [inline hubble-proxy/cmd/root.go:31]: why?

File: hubble-proxy/cmd/version/version.go
Comment (glibsm) [inline hubble-proxy/cmd/version/version.go]: there is an stdout attached to the cmd object, rather than os (useful to get into the habbit of that for future testing implications)

**Review by glibsm — APPROVED** [review]
As a skeleton looks good. The remarks are not blocking but are better saved for the upcoming PRs

File: hubble-proxy/cmd/completion/completion.go
Comment (tgraf) [reply hubble-proxy/cmd/completion/completion.go:1]: We should discuss this broader in #development on Slack and then apply it universally.

**Comment by tgraf** [comment]
@joestringer @glibsm Reading through the review comments, it wasn't clear to me whether we should merge this now and follow-up or require some of the comments to be addressed now. I assume Robin will look through and address what's obvious, it would be great for both of you to point out what you want to be addressed now vs later after Robin has refreshed the PR.

File: hubble-proxy/cmd/root.go
Comment (rolinh) [reply hubble-proxy/cmd/root.go:31]: By default, usage is always being displayed when a command returns an error, whether that's a user input issue or any other error. It'd be OK to display it when a command is called with invalid arguments but it feels rather odd to display it when it's an internal error such as a connection timeout or else.

File: hubble-proxy/cmd/completion/completion.go
Comment (rolinh) [reply hubble-proxy/cmd/completion/completion.go:1]: Indeed, I think this should be tackled as part of #9846.

**Comment by Rolinh** [comment]
test-me-please

**Comment by joestringer** [comment]
The specific items I was concerned about are already addressed.

I had a minor longer-term maintenance question on refactoring the basic cmd code here to better share, which I raised because this PR fixed an issue in the Cilium side as well and we just applied the same fix there:
https://github.com/cilium/cilium/pull/10545#discussion_r391356975

That said, the relevant code doesn't change often so it's not a bit deal to have multiple copies of it in the codebase.

**Pull request merged**
